// @aws-cdk/aws-kinesisfirehose-destinations-alpha/lib/s3-bucket.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3Bucket = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("aws-cdk-lib/aws-iam");
const common_1 = require("./common");
const helpers_1 = require("./private/helpers");
/**
 * An S3 bucket destination for data from a Kinesis Data Firehose delivery stream.
 */
class S3Bucket {
    constructor(bucket, props = {}) {
        var _b;
        this.bucket = bucket;
        this.props = props;
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_kinesisfirehose_destinations_alpha_S3BucketProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.constructor);
            }
            throw error;
        }
        if (((_b = this.props.s3Backup) === null || _b === void 0 ? void 0 : _b.mode) === common_1.BackupMode.FAILED) {
            throw new Error('S3 destinations do not support BackupMode.FAILED');
        }
    }
    bind(scope, _options) {
        var _b, _c, _d, _e;
        const role = (_b = this.props.role) !== null && _b !== void 0 ? _b : new iam.Role(scope, 'S3 Destination Role', {
            assumedBy: new iam.ServicePrincipal('firehose.amazonaws.com'),
        });
        const bucketGrant = this.bucket.grantReadWrite(role);
        const { loggingOptions, dependables: loggingDependables } = (_c = helpers_1.createLoggingOptions(scope, {
            logging: this.props.logging,
            logGroup: this.props.logGroup,
            role,
            streamId: 'S3Destination',
        })) !== null && _c !== void 0 ? _c : {};
        const { backupConfig, dependables: backupDependables } = (_d = helpers_1.createBackupConfig(scope, role, this.props.s3Backup)) !== null && _d !== void 0 ? _d : {};
        return {
            extendedS3DestinationConfiguration: {
                cloudWatchLoggingOptions: loggingOptions,
                processingConfiguration: helpers_1.createProcessingConfig(scope, role, this.props.processor),
                roleArn: role.roleArn,
                s3BackupConfiguration: backupConfig,
                s3BackupMode: this.getS3BackupMode(),
                bufferingHints: helpers_1.createBufferingHints(this.props.bufferingInterval, this.props.bufferingSize),
                bucketArn: this.bucket.bucketArn,
                compressionFormat: (_e = this.props.compression) === null || _e === void 0 ? void 0 : _e.value,
                encryptionConfiguration: helpers_1.createEncryptionConfig(role, this.props.encryptionKey),
                errorOutputPrefix: this.props.errorOutputPrefix,
                prefix: this.props.dataOutputPrefix,
            },
            dependables: [bucketGrant, ...(loggingDependables !== null && loggingDependables !== void 0 ? loggingDependables : []), ...(backupDependables !== null && backupDependables !== void 0 ? backupDependables : [])],
        };
    }
    getS3BackupMode() {
        var _b, _c;
        return ((_b = this.props.s3Backup) === null || _b === void 0 ? void 0 : _b.bucket) || ((_c = this.props.s3Backup) === null || _c === void 0 ? void 0 : _c.mode) === common_1.BackupMode.ALL
            ? 'Enabled'
            : undefined;
    }
}
exports.S3Bucket = S3Bucket;
_a = JSII_RTTI_SYMBOL_1;
S3Bucket[_a] = { fqn: "@aws-cdk/aws-kinesisfirehose-destinations-alpha.S3Bucket", version: "2.23.0-alpha.0" };
